internal/signinwithssb/challenges.go

```diff
@@ -0,0 +1,76 @@
+// SPDX-License-Identifier: MIT
+
+package signinwithssb
+
+import (
+	"bytes"
+	"crypto/rand"
+	"encoding/base64"
+	"fmt"
+
+	"golang.org/x/crypto/ed25519"
+
+	refs "go.mindeco.de/ssb-refs"
+)
+
+// sign-in with ssb uses 256-bit nonces
+const challengeLength = 32
+
+// DecodeChallengeString accepts base64 encoded strings and decodes them,
+// checks their length to be equal to challengeLength,
+// and returns the decoded bytes
+func DecodeChallengeString(c string) ([]byte, error) {
+	challengeBytes, err := base64.URLEncoding.DecodeString(c)
+	if err != nil {
+		return nil, fmt.Errorf("invalid challenge encoding: %w", err)
+	}
+
+	if n := len(challengeBytes); n != challengeLength {
+		return nil, fmt.Errorf("invalid challenge length: expected %d but got %d", challengeLength, n)
+	}
+
+	return challengeBytes, nil
+}
+
+// GenerateChallenge returs a base64 encoded string
+//  with challangeLength bytes of random data
+func GenerateChallenge() string {
+	buf := make([]byte, challengeLength)
+	rand.Read(buf)
+	return base64.URLEncoding.EncodeToString(buf)
+}
+
+// ClientRequest is used to verify an incoming client response
```

Comment:
Not very important to change, but perhaps for other devs reading this code, they may be confused by "request" when this is in reality a response. Or which one is it?

Reply:
True.. It can be both, actually.

There is `Sign(privateKey) (signature []bytes)` and `Verify(signature []byte) bool` on it.

I could do a similar thing to what I did on aliases here.

Reply:
I guess `ClientChallenge` would be more generic then `Request` but it's members are also challenges and it felt odd.

Reply:
ClientPayload?